### PR TITLE
The promised leftovers from the previous PR

### PR DESCRIPTION
**Pull request by acecilia**
Let me know if I missed something or you prefer other names

### CODE REVIEW

**Comment by segiddins** [comment]
I'm actually having trouble running tests locally now.

```
$ bazelisk test --local_test_jobs=1 --apple_platform_type=ios //tests/ios/... -k

INFO: Call stack for the definition of repository 'carthage' which is a unexported__carthage_impl (rule definition at /Users/segiddins/Development/OpenSource/rules_ios/rules/framework_builder.bzl:172:36):
 - <builtin>
 - /Users/segiddins/Development/OpenSource/rules_ios/rules/framework_builder.bzl:191:5
 - /Users/segiddins/Development/OpenSource/rules_ios/rules/framework_builder.bzl:78:5
 - /Users/segiddins/Development/OpenSource/rules_ios/WORKSPACE:60:1
ERROR: An error occurred during the fetch of repository 'carthage':
   Traceback (most recent call last):
        File "/Users/segiddins/Development/OpenSource/rules_ios/rules/framework_builder.bzl", line 92
                _execute(ctx, ctx.attr.cmd)
        File "/Users/segiddins/Development/OpenSource/rules_ios/rules/framework_builder.bzl", line 44, in _execute
                fail(<1 more arguments>)

====== Command 'carthage bootstrap --no-use-binaries --platform iOS' failed (127) ======
====== stdout ======

====== stderr ======
sh: carthage: command not found

        
ERROR: /Users/segiddins/Development/OpenSource/rules_ios/tests/ios/frameworks/mixed-source/sources-with-prebuilt-binaries/BUILD.bazel:5:1: //tests/ios/frameworks/mixed-source/sources-with-prebuilt-binaries:_InputMask depends on @carthage//:InputMask in repository @carthage which failed to fetch. no such package '@carthage//': Traceback (most recent call last):
        File "/Users/segiddins/Development/OpenSource/rules_ios/rules/framework_builder.bzl", line 92
                _execute(ctx, ctx.attr.cmd)
        File "/Users/segiddins/Development/OpenSource/rules_ios/rules/framework_builder.bzl", line 44, in _execute
                fail(<1 more arguments>)

====== Command 'carthage bootstrap --no-use-binaries --platform iOS' failed (127) ======
====== stdout ======

====== stderr ======
sh: carthage: command not found

        
INFO: Call stack for the definition of repository 'cocoapods' which is a unexported__cocoapods_impl (rule definition at /Users/segiddins/Development/OpenSource/rules_ios/rules/framework_builder.bzl:172:36):
 - <builtin>
 - /Users/segiddins/Development/OpenSource/rules_ios/rules/framework_builder.bzl:191:5
 - /Users/segiddins/Development/OpenSource/rules_ios/rules/framework_builder.bzl:131:5
 - /Users/segiddins/Development/OpenSource/rules_ios/WORKSPACE:65:1
ERROR: An error occurred during the fetch of repository 'cocoapods':
   Traceback (most recent call last):
        File "/Users/segiddins/Development/OpenSource/rules_ios/rules/framework_builder.bzl", line 148
                _execute(ctx, ctx.attr.cmd)
        File "/Users/segiddins/Development/OpenSource/rules_ios/rules/framework_builder.bzl", line 44, in _execute
                fail(<1 more arguments>)

====== Command 'bundle install && bundle exec pod install' failed (1) ======
====== stdout ======
Fetching gem metadata from https://rubygems.org/..........
Resolving dependencies...
Fetching CFPropertyList 3.0.2
Installing CFPropertyList 3.0.2
Fetching concurrent-ruby 1.1.6
Installing concurrent-ruby 1.1.6
Fetching i18n 0.9.5
Installing i18n 0.9.5
Fetching minitest 5.14.1
Installing minitest 5.14.1
Fetching thread_safe 0.3.6
Installing thread_safe 0.3.6
Fetching tzinfo 1.2.7
Installing tzinfo 1.2.7
Fetching activesupport 4.2.11.3
Installing activesupport 4.2.11.3
Fetching httpclient 2.8.3
Installing httpclient 2.8.3
Fetching json 2.3.0
Installing json 2.3.0 with native extensions
Fetching algoliasearch 1.27.2
Installing algoliasearch 1.27.2
Fetching atomos 0.1.3
Installing atomos 0.1.3
Using bundler 2.1.4
Fetching claide 1.0.3
Installing claide 1.0.3
Fetching fuzzy_match 2.0.4
Installing fuzzy_match 2.0.4
Fetching nap 1.1.0
Installing nap 1.1.0
Fetching cocoapods-core 1.8.4
Installing cocoapods-core 1.8.4
Fetching cocoapods-deintegrate 1.0.4
Installing cocoapods-deintegrate 1.0.4
Fetching cocoapods-downloader 1.3.0
Installing cocoapods-downloader 1.3.0
Fetching cocoapods-plugins 1.0.0
Installing cocoapods-plugins 1.0.0
Fetching cocoapods-search 1.0.0
Installing cocoapods-search 1.0.0
Fetching cocoapods-stats 1.1.0
Installing cocoapods-stats 1.1.0
Fetching netrc 0.11.0
Installing netrc 0.11.0
Fetching cocoapods-trunk 1.5.0
Installing cocoapods-trunk 1.5.0
Fetching cocoapods-try 1.2.0
Installing cocoapods-try 1.2.0
Fetching colored2 3.1.2
Installing colored2 3.1.2
Fetching escape 0.0.4
Installing escape 0.0.4
Fetching fourflusher 2.3.1
Installing fourflusher 2.3.1
Fetching gh_inspector 1.1.3
Installing gh_inspector 1.1.3
Fetching molinillo 0.6.6
Installing molinillo 0.6.6
Fetching ruby-macho 1.4.0
Installing ruby-macho 1.4.0
Fetching nanaimo 0.2.6
Installing nanaimo 0.2.6
Fetching xcodeproj 1.16.0
Installing xcodeproj 1.16.0
Fetching cocoapods 1.8.4
Installing cocoapods 1.8.4
Fetching rouge 2.0.7
Installing rouge 2.0.7
Fetching xcpretty 0.3.0
Installing xcpretty 0.3.0
Fetching cocoapods-binary 0.4.4
Installing cocoapods-binary 0.4.4
Bundle complete! 2 Gemfile dependencies, 36 gems now installed.
Bundled gems are installed into `./vendor/bundle`
🚀  Prebuild frameworks
Analyzing dependencies
Downloading dependencies
Installing SnapKit (5.0.1)
Generating Pods project
Skipping User Project Integration
Prebuild frameworks (total 1)
Prebuilding SnapKit...

====== stderr ======

[!] The abstract target Pods is not inherited by a concrete target, so the following dependencies won't make it into any targets in your project:
    - SnapKit (= 5.0.1)

        
ERROR: /Users/segiddins/Development/OpenSource/rules_ios/tests/ios/frameworks/mixed-source/sources-with-prebuilt-binaries/BUILD.bazel:11:1: //tests/ios/frameworks/mixed-source/sources-with-prebuilt-binaries:_SnapKit depends on @cocoapods//:SnapKit in repository @cocoapods which failed to fetch. no such package '@cocoapods//': Traceback (most recent call last):
        File "/Users/segiddins/Development/OpenSource/rules_ios/rules/framework_builder.bzl", line 148
                _execute(ctx, ctx.attr.cmd)
        File "/Users/segiddins/Development/OpenSource/rules_ios/rules/framework_builder.bzl", line 44, in _execute
                fail(<1 more arguments>)

====== Command 'bundle install && bundle exec pod install' failed (1) ======
====== stdout ======
Fetching gem metadata from https://rubygems.org/..........
Resolving dependencies...
Fetching CFPropertyList 3.0.2
Installing CFPropertyList 3.0.2
Fetching concurrent-ruby 1.1.6
Installing concurrent-ruby 1.1.6
Fetching i18n 0.9.5
Installing i18n 0.9.5
Fetching minitest 5.14.1
Installing minitest 5.14.1
Fetching thread_safe 0.3.6
Installing thread_safe 0.3.6
Fetching tzinfo 1.2.7
Installing tzinfo 1.2.7
Fetching activesupport 4.2.11.3
Installing activesupport 4.2.11.3
Fetching httpclient 2.8.3
Installing httpclient 2.8.3
Fetching json 2.3.0
Installing json 2.3.0 with native extensions
Fetching algoliasearch 1.27.2
Installing algoliasearch 1.27.2
Fetching atomos 0.1.3
Installing atomos 0.1.3
Using bundler 2.1.4
Fetching claide 1.0.3
Installing claide 1.0.3
Fetching fuzzy_match 2.0.4
Installing fuzzy_match 2.0.4
Fetching nap 1.1.0
Installing nap 1.1.0
Fetching cocoapods-core 1.8.4
Installing cocoapods-core 1.8.4
Fetching cocoapods-deintegrate 1.0.4
Installing cocoapods-deintegrate 1.0.4
Fetching cocoapods-downloader 1.3.0
Installing cocoapods-downloader 1.3.0
Fetching cocoapods-plugins 1.0.0
Installing cocoapods-plugins 1.0.0
Fetching cocoapods-search 1.0.0
Installing cocoapods-search 1.0.0
Fetching cocoapods-stats 1.1.0
Installing cocoapods-stats 1.1.0
Fetching netrc 0.11.0
Installing netrc 0.11.0
Fetching cocoapods-trunk 1.5.0
Installing cocoapods-trunk 1.5.0
Fetching cocoapods-try 1.2.0
Installing cocoapods-try 1.2.0
Fetching colored2 3.1.2
Installing colored2 3.1.2
Fetching escape 0.0.4
Installing escape 0.0.4
Fetching fourflusher 2.3.1
Installing fourflusher 2.3.1
Fetching gh_inspector 1.1.3
Installing gh_inspector 1.1.3
Fetching molinillo 0.6.6
Installing molinillo 0.6.6
Fetching ruby-macho 1.4.0
Installing ruby-macho 1.4.0
Fetching nanaimo 0.2.6
Installing nanaimo 0.2.6
Fetching xcodeproj 1.16.0
Installing xcodeproj 1.16.0
Fetching cocoapods 1.8.4
Installing cocoapods 1.8.4
Fetching rouge 2.0.7
Installing rouge 2.0.7
Fetching xcpretty 0.3.0
Installing xcpretty 0.3.0
Fetching cocoapods-binary 0.4.4
Installing cocoapods-binary 0.4.4
Bundle complete! 2 Gemfile dependencies, 36 gems now installed.
Bundled gems are installed into `./vendor/bundle`
🚀  Prebuild frameworks
Analyzing dependencies
Downloading dependencies
Installing SnapKit (5.0.1)
Generating Pods project
Skipping User Project Integration
Prebuild frameworks (total 1)
Prebuilding SnapKit...

====== stderr ======

[!] The abstract target Pods is not inherited by a concrete target, so the following dependencies won't make it into any targets in your project:
    - SnapKit (= 5.0.1)
```

**Comment by ob** [comment]
Hmm, I thought the problem was just that you were missing `carthage`  but after running `brew install carthage` on my machine it stalls for more than 2 minutes fetching `@cocoapods` and `@carthage` and fails with

```
ERROR: /Users/obonilla/o/rules_ios/tests/ios/frameworks/mixed-source/sources-with-prebuilt-binaries/BUILD.bazel:17:1: Compiling Swift module MixedSourceFramework failed (Exit 1): worker failed: error executing command
  (cd /private/var/tmp/_bazel_obonilla/ec95186ea51d22485ad49f25cb412589/execroot/build_bazel_rules_ios && \
  exec env - \
    APPLE_SDK_PLATFORM=MacOSX \
    APPLE_SDK_VERSION_OVERRIDE=10.15 \
    XCODE_VERSION_OVERRIDE=11.3.1.11C505 \
  bazel-out/host/bin/external/build_bazel_rules_swift/tools/worker/worker swiftc @bazel-out/apl-darwin_x86_64-fastbuild/bin/tests/ios/frameworks/mixed-source/sources-with-prebuilt-binaries/MixedSourceFramework.swiftmodule-0.params)
Execution platform: @local_config_platform//:host
tests/ios/frameworks/mixed-source/sources-with-prebuilt-binaries/MixedSourceFramework/Framework.swift:1:8: error: module 'InputMask' was created for incompatible target x86_64-apple-ios8.0-simulator: /private/var/tmp/_bazel_obonilla/ec95186ea51d22485ad49f25cb412589/execroot/build_bazel_rules_ios/external/carthage/Carthage/Build/iOS/InputMask.framework/Modules/InputMask.swiftmodule/x86_64.swiftmodule
import InputMask
       ^

INFO: Elapsed time: 197.701s, Critical Path: 45.92s
INFO: 690 processes: 602 darwin-sandbox, 78 local, 10 worker.
FAILED: Build did NOT complete successfully
```

I'm thinking we should revert while we investigate why CI passed when this fails locally and figure out how to minimize the impact at loading time. Thoughts @acecilia, @segiddins ?

**Comment by segiddins** [comment]
Maybe we keep the rules changes but revert the repository rules ?

**Comment by ob** [comment]
We could also split the bazel workspace in two (add a different workspace file) and put the repository rules there so that you can build and test the rules without the repository rules. Thinking out loud here...

**Comment by acecilia** [comment]
@segiddins Yep, you are missing `carthage` in your system. Some rationale about the process:

* When installing `cocoapods` we are using your `gemfile` to install it: if you have `bundle` installed in your system (which in macos comes installed by default), you are good to go. This is because `bundle` is the recommended way of installing a pinned version of `cocoapods`. 
* When installing `carthage` there is no recommended way of installing a pinned version. Usually I would use `mint` for it, but it is not a widespread method, and `mint` does not come preinstalled neither in macos or the github build agents.

@ob I have seen your output before when I forget to add `--apple_platform_type=ios` to the bazel command. Let me know if that is the problem

**Comment by segiddins** [comment]
I shouldn’t be missing anything, the rules should be downloading everything they need in order for me to run `bazel build`

**Comment by acecilia** [comment]
@segiddins that is not a fair comment after my explanation, and is definitely not constructive feedback. 

If you are using a rule that uses carthage to build binaries, then you need carthage in your system. I do not see a way around that. If you want to automate installing carthage in the host system, that is a different more complex issue: it is something that a package manager does.

If you have any other constructive proposal about how to solve this I am open to suggestions, and will invest the time and effort needed because I want this working as good as possible.

**Comment by ob** [comment]
Doesn't Carthage have a binary distribution? Seems like the repository rules could add it as a dependency and bazel could download it on demand no?

**Comment by acecilia** [comment]
Thanks for the suggestion @ob. I am looking into it :) Did it build for you with `--apple_platform_type=ios`?

**Comment by ob** [comment]
I'll play with it tonight, I have an idea for not having that long analysis pause. I think we can move `framework_builder.bzl` to a new subdir (`repository_rules`) that sits next to `rules` and move the test in `sources-with-prebuilt-binaries/MixedSourceFramework` to a different package (i.e. a subdirectory of `tests` with its own `WORKSPACE` file). Then you only hit that loading stuff when you are working on the repository rules.

**Comment by acecilia** [comment]
@ob I fixed the complains of @segiddins about carthage, see last commit of this PR.
I will amend this PR when you merge yours.

**Comment by ob** [comment]
Cool, I merged my PR but had to revert the multiple `WORKSPACE` part... it turns out Bazel won't stop at `WORKSPACE` boundaries so it still loads all the `BUILD` files in the repo.

**Comment by acecilia** [comment]
Pitty... Well, it should not be that bad, as when you build it once it gets cached forever. I will see if I can come up with something.

This PR is ready on green :)

**Comment by acecilia** [comment]
Maybe this is the way to go, ignore the tests locally and force building them on CI: https://github.com/bazelbuild/bazel/issues/2460#issuecomment-422009106

File: .bazelrc [hunk 1]
```diff
@@ -3,3 +3,7 @@
 build --strategy=Stardoc=standalone
 build --verbose_failures # Print the full command line for commands that failed
 build --test_output=errors # Prints log file output to the console on failure
+
+# By default do not build the tests for sources-with-prebuilt-binaries, 
+# because it takes quiet some time. They will only run on CI
```
Comment (ob) [inline .bazelrc:8]:
```suggestion
# because it takes quite some time. They will only run on CI
```

**Pull request merged**